### PR TITLE
issue[#395]Optimizer.optimize update and return the input model

### PR DESCRIPTION
**Pull request by qiuxin2012**
#395 
Add two functions to nn.Utils, copyModule and recursiveCopyState.
Use copyModule at the end of DistriOptimizer.optimize to copy the trained Module's parameters and running state to `model`.
Use recursiveCopyState at the end of LocalOptimizer.optimize to copy the `workingModels`'s running state to the `model`, and parameters is already shared between `workingModels` and `model`.

### CODE REVIEW

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 1]
```diff
@@ -461,10 +461,10 @@ object DistriOptimizer {
...
   }
 
-
-  private def getModel[T: ClassTag](models: RDD[Cache[T]],
-      parameters: AllReduceParameter[T])
-  : Module[T] = {
+  private def getModel[T: ClassTag](
+      models: RDD[Cache[T]],
+      parameters: AllReduceParameter[T],
+      result: Module[T] = null): Module[T] = {
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: There is no need to change the interface; you can unconditionally copy the model you get in this method to the user provided model (Optimizer.model), either at the end of this method or at the end of the optimize() method.

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 2]
```diff
@@ -485,7 +485,17 @@ object DistriOptimizer {
...
+      // Copy trainedModel to result
+      val resultParameter = result.getParameters()._1
+      require(resultParameter.nElement() == parameter.nElement(),
+        "result's parameter should have the same size with train model's parameter")
+      resultParameter.copy(parameter)
+      nn.Utils.recursiveCopyState(trainedModel, result)
+      result
+    } else {
+      trainedModel
+    }
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: You can just return trainedModel unconditionally here

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 3]
```diff
@@ -485,7 +485,17 @@ object DistriOptimizer {
       parameter.narrow(1, start + 1, length).copy(weights(pid))
     })
 
-    trainedModel
+    if (null != result) {
+      // Copy trainedModel to result
+      val resultParameter = result.getParameters()._1
+      require(resultParameter.nElement() == parameter.nElement(),
+        "result's parameter should have the same size with train model's parameter")
+      resultParameter.copy(parameter)
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: resultParameter.set(parameter)

File: dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 4]
```diff
@@ -485,7 +485,17 @@ object DistriOptimizer {
...
+      nn.Utils.recursiveCopyState(trainedModel, result)
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: I think you can define a recursiveCopyModule method, which in turn takes care of both the parameter setting and state copy.

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala [hunk 5]
```diff
@@ -218,4 +218,32 @@ object Utils {
...
+        val c2 = result.asInstanceOf[Container[Activity, Activity, T]]
+        require(c1.modules.length == c2.modules.length)
+        var i = 0
+        while (i < c1.modules.length) {
+          recursiveCopyState(c1.modules(i), c2.modules(i))
+          i += 1
+        }
+      }
+    }
+  }
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala]: Why not
```scala
if (model.isInstanceOf[Container[Activity, Activity, T]]) {
   //report error if result is not a container
   ...
}
else if (model.isInstanceOf[BatchNormalization[T]]) {
   //report error if result is not a BatchNormalization
   ...
}
```

**Comment by jason-dai** [comment]
You should copy state for the local optimizer too

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala [hunk 6]
```diff
@@ -218,4 +218,51 @@ object Utils {
...
+      val b1 = src.asInstanceOf[BatchNormalization[T]]
+      val b2 = dst.asInstanceOf[BatchNormalization[T]]
+      b2.runningMean.copy(b1.runningMean)
+      b2.runningVar.copy(b1.runningVar)
+    } else if (src.isInstanceOf[Container[Activity, Activity, T]]) {
+      require(dst.isInstanceOf[Container[Activity, Activity, T]],
+        s"$dst must be the same type with $src")
+      val c1 = src.asInstanceOf[Container[Activity, Activity, T]]
+      val c2 = dst.asInstanceOf[Container[Activity, Activity, T]]
+      require(c1.modules.length == c2.modules.length)
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala]: require message?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala [hunk 7]
```diff
@@ -218,4 +218,51 @@ object Utils {
...
+   */
+  def copyModule[T](src: Module[T], dst: Module[T]): Module[T] = {
+    // copy parameters
+    val srcParameters = src.getParameters()._1
+    val dstParameters = dst.getParameters()._1
+    require(srcParameters.size(1) == dstParameters.size(1),
+      s"$src and $dst is not the same type.")
+    dstParameters.copy(srcParameters)
+    // copy state
+    if (src.isInstanceOf[Container[Activity, Activity, T]] ||
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala]: do we still need the if here?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala [hunk 8]
```diff
@@ -218,4 +218,51 @@ object Utils {
...
+    }
+    dst
+  }
+
+  /**
+   * copy runningMean and runningVar of BatchNormalization into result recursively
+   * @param src
+   * @param dst
+   */
+  def recursiveCopyState[T](src: Module[T], dst: Module[T]): Unit = {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala]: Can we move this into module memeber?

**Comment by yiheng** [comment]
LGTM

**Comment by qiuxin2012** [comment]
Some update.
Add a module's method `copyStatus()` to copy module's status instead of `recursiveCopyState()`.

**Pull request merged**